blocks/serialization.py

```diff
@@ -1,83 +1,109 @@
-reall need to do this, it is often possible to write a wrapper function such \
-as `def lower(s): return s.lower()` and store that instead."""
+class PersistentParameterID(PersistentSharedVariableID):
+    """Persist the names of parameter arrays in the zip file.
 
-LAMBDA_ERROR = """
-
-Python is unable to serialize lambda functions. Make sure that your code does \
-not rely on them and use normal function definitions instead."""
-
-NESTED_FUNCTION_ERROR = """
-
-Python is unable to serialize nested functions. Make sure that all your \
-functions are available in the global namespace."""
-
-
-def pickle_dump(*args, **kwargs):
-    """A wrapper around pickle's dump that provides informative errors."""
-    kwargs.setdefault('protocol', DEFAULT_PROTOCOL)
-    try:
-        cPickle.dump(*args, **kwargs)
-    except Exception as e:
-        if six.PY3 and '<lambda>' in e.args[0]:
-            reraise_as("Pickling failed to pickle a lambda function." +
-                       LAMBDA_ERROR)
-        if six.PY3 and '<function' in e.args[0] and '<locals>' in e.args[0]:
-            reraise_as("Pickling failed to pickle a nested function." +
-                       NESTED_FUNCTION_ERROR)
-        if six.PY2 and 'function objects' in e.args[0]:
-            reraise_as("Pickling failed to pickle a function." +
-                       LAMBDA_ERROR + NESTED_FUNCTION_ERROR)
-        if ((six.PY2 and 'isinstancemethod' in e.args[0]) or
-                (six.PY3 and '<function' in e.args[0] and
-                 'attribute lookup' in e.args[0])):
-            reraise_as("Pickling failed to pickle a reference to a method." +
-                       INSTANCEMETHOD_ERROR)
-        reraise_as("Pickling failed." + PICKLING_ERROR)
+    If a shared variable has a name, this name is used as the name of the
+    NPY file inside of the zip file. NumPy arrays that aren't matched to a
+    shared variable are persisted as usual (i.e. `array_0`, `array_1`,
+    etc.)
 
+    Parameters
+    ----------
+    allow_unnamed : bool, optional
+        Allow shared variables without a name to be persisted. Defaults to
+        ``True``.
+    allow_duplicates : bool, optional
+        Allow multiple shared variables to have the same name, in which
+        case they will be numbered e.g. `x`, `x_2`, `x_3`, etc. Defaults to
+        ``True``.
+
+    Raises
+    ------
+    ValueError
+        If an unnamed shared variable is encountered and `allow_unnamed` is
+        ``False``, or if two shared variables have the same name, and
+        `allow_duplicates` is ``False``.
 
-def secure_pickle_dump(object_, path):
-    """Robust serialization - does not corrupt your files when failed.
+    """
+    def __call__(self, obj):
+        if isinstance(obj, SharedVariable):
+            if obj.name:
+                if obj.name == 'pkl':
+                    ValueError("can't pickle shared variable with name `pkl`")
+                if hasattr(obj, 'tag'):
```

Comment:
Don't shared variables always have the `tag` attribute? It could be an empty dictionary though if it's not annotated.

Thinking about it, this class should probably inherit from `PersistentNdarrayID`. Right now, when `super` gets called, doesn't the parent class (`PersistentSharedVariableID`) simply override the name you've just created?